ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,461 @@
+            // precomputation for first proof
+            // first denominator
+            let first_denominator = CanonicalLagrangeDenominator::<
+                Fp61BitPrime,
+                { LargeProofGenerator::RECURSION_FACTOR },
+            >::new();
+            // first lagrange table
+            let first_lagrange_table = LagrangeTable::<
+                Fp61BitPrime,
+                { LargeProofGenerator::RECURSION_FACTOR },
+                { LargeProofGenerator::LAGRANGE_LENGTH },
+            >::from(first_denominator);
+
+            LargeProofGenerator::gen_artefacts_from_recursive_step(
+                ctx,
+                &mut record_counter,
+                &first_lagrange_table,
+                Self::polynomials_from_blocks(uv_tuple_block),
+            )
+        };
+
+        // storage for other proofs
+        let mut my_proofs_left_shares =
+            Vec::<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>::new();
+        let mut shares_of_proofs_from_prover_left =
+            Vec::<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>::new();
+        // and masks
+        let mut p_mask_from_left_prover = Fp61BitPrime::ZERO;
+        let mut q_mask_from_right_prover = Fp61BitPrime::ZERO;
+
+        // precomputation for other proofs
+        // denominator
+        let denominator = CanonicalLagrangeDenominator::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+        >::new();
+        // lagrange table
+        let lagrange_table = LagrangeTable::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+            { SmallProofGenerator::LAGRANGE_LENGTH },
+        >::from(denominator);
+
+        // recursively generate proofs
+        // via SmallProofGenerator
+        while uv_values.len() > 1 {
+            if uv_values.len() < SmallProofGenerator::RECURSION_FACTOR {
+                // generate masks
+                // verifier on the right has p,
+                // therefore the right share is "implicitly sent" to the right ("communicated" via PRSS)
+                (p_mask_from_left_prover, q_mask_from_right_prover) =
+                    uv_values.set_masks(ctx, &mut record_counter).unwrap();
+            }
+
+            // generate next proof
+            let (uv_values_new, share_of_proof_from_prover_left, my_proof_left_share) =
+                SmallProofGenerator::gen_artefacts_from_recursive_step(
+                    ctx,
+                    &mut record_counter,
+                    &lagrange_table,
+                    uv_values.iter(),
+                );
+            // collect proof
+            shares_of_proofs_from_prover_left.push(share_of_proof_from_prover_left);
+            my_proofs_left_shares.push(my_proof_left_share);
+
+            uv_values = uv_values_new;
+        }
+
+        // output proofs
+        // and masks
+        (
```

Comment:
I think this return type is probably too complex for a tuple. We should use a structure so that we don't have to remember what .0 and .1 and .2 and .3 all refer to.

Reply:
Actually... we should just inline this function. It is only called in a single place. Then we don't have to worry about making any new structures, or thinking about complex return types.